### ISSUE TITLE
bug: Controller Modals are not accessible on iOS (voiceOver) 

### ISSUE THREAD

**Issue by sanaadostmuhammad** [ionic-team/ionic-framework#26872]
### Prerequisites

- [X] I have read the [Contributing Guidelines](https://github.com/ionic-team/ionic-framework/blob/main/.github/CONTRIBUTING.md#creating-an-issue).
- [X] I agree to follow the [Code of Conduct](https://ionicframework.com/code-of-conduct).
- [X] I have searched for [existing issues](https://github.com/ionic-team/ionic-framework/issues) that already report this problem, without success.

### Ionic Framework Version

- [ ] v4.x
- [ ] v5.x
- [X] v6.x
- [ ] v7.x
- [ ] Nightly

### Current Behavior

Controller modals are not being read by voice over (iOS). 
When a modal is launched with a button click, the voice over focus stays on the button. The screen content is rendered but not being detected by the voice over. Even the 'Two-finger swipe up' gesture isn't helping. 

### Expected Behavior

Voice over should be able to read content of the modal. 

### Steps to Reproduce

In the linked repo, we are using the official documentation controller modal example. Click on the first 'OPEN' button and you will notice that voice over is stuck on the button. It would neither read the underlying page nor the modal content.

If you try to read the entire screen from top using 'Two-finger swipe up', you will notice that only the phone status bar is being read. 

### Code Reproduction URL

https://github.com/albiorixvimal/ionic6-modal

### Ionic Info

Ionic:

   Ionic CLI                     : 6.20.3 (/usr/local/lib/node_modules/@ionic/cli)
   Ionic Framework               : @ionic/angular 6.5.6
   @angular-devkit/build-angular : 15.1.2
   @angular-devkit/schematics    : 15.1.2
   @angular/cli                  : 15.1.2
   @ionic/angular-toolkit        : 6.1.0

Cordova:

   Cordova CLI       : 9.0.0 (cordova-lib@9.0.1)
   Cordova Platforms : android 10.1.2, ios 6.2.0
   Cordova Plugins   : cordova-plugin-ionic-keyboard 2.2.0, cordova-plugin-ionic-webview 5.0.0, (and 3 other plugins)

Utility:

   cordova-res (update available: 0.15.4) : 0.15.3
   native-run (update available: 1.7.1)   : 0.3.0

System:

   Android SDK Tools : 26.1.1 (/Users/sanaamuhammad/Library/Android/sdk)
   ios-deploy        : 1.9.4
   ios-sim           : 8.0.2
   NodeJS            : v16.19.0 (/usr/local/bin/node)
   npm               : 8.19.3
   OS                : macOS Monterey
   Xcode             : Xcode 14.2 Build version 14C18

**Comment by liamdebeasi** [comment]
Thanks for the report. I am unable to reproduce this behavior on my iOS 16 device. What version of iOS are you using to test this? Also what environment does this reproduce in? (PWA, Cordova, Capacitor, Browser, etc)

https://user-images.githubusercontent.com/2721089/221876028-facc2b5d-5d0d-4295-ba5b-d04e314cbc64.mp4

**Comment by sanaadostmuhammad** [comment]
https://user-images.githubusercontent.com/24274634/222047944-73b3fc08-8cc5-4ac1-a467-904298bf5a9e.mov

The sample project is tested on iOS 15.4.1 with Cordova iOS 6.2.0

**Comment by sanaadostmuhammad** [comment]
@liamdebeasi, I just noticed that you are testing the inline modal and not the **Controller modal**. The accessibility issue is with the controller modal.

**Comment by liamdebeasi** [comment]
Your reproduction case only includes an inline modal test case which is why I tested that. Can you please provide a controller modal test case?

**Comment by sanaadostmuhammad** [comment]
My bad, I have pushed the latest code. 

Can be tested on Ionic site too. https://ionicframework.com/docs/api/modal#controller-modals 
https://user-images.githubusercontent.com/24274634/222291806-e137b6b6-56d6-409c-a773-61b761309491.MP4

**Comment by liamdebeasi** [comment]
Thanks for the reply. It looks like the demo on https://ionicframework.com/docs/api/modal#controller-modals was written incorrectly which is why focus trapping does not work as expected. I'll fix that demo. However I am unable to reproduce the issue in your sample repo. Do you have a video of this issue reproducing in your sample app?

**Comment by sanaadostmuhammad** [comment]
The following video is of the sample app.
https://github.com/ionic-team/ionic-framework/issues/26872#issuecomment-1449344040

Can you please explain the problem with the demo? Our offical app has 50+ controller modals and not all of them can be replaced with inline modals. 

This bug is blocking us from migrating to ionic 6 as our visually impaired customers rely on voiceOver.

**Comment by liamdebeasi** [comment]
I'm still not able to reproduce this issue in your application. I tested in both iOS 15.6 and iOS 16.4 and focus is correctly moved to the modal. Here is a demo on iOS 15.6:

https://user-images.githubusercontent.com/2721089/222456755-e5c70516-7734-49b6-ac46-c7f00d75a3fd.mp4

Does this issue happen if you access the app from a web browser? Or does it only reproduce in Cordova?


-----

For more context, Ionic will manually move focus inside of the modal when it is presented. We also hide contents underneath the modal using `aria-hidden` so it is not accidentally read by screen readers. We hide contents underneath the modal by adding an `aria-hidden` to the `ion-router-outlet` component. Our playground examples do not use `ion-router-outlet`, which is why focus trapping is not working there. However, your application does use `ion-router-outlet` so content underneath the modal is being hidden.

**Comment by liamdebeasi** [comment]
I was able to get a device running iOS 15.4. It looks like this is a bug in Safari/WebKit that is unique to that version. Let me see if there is a workaround.

**Comment by sanaadostmuhammad** [comment]
> I'm still not able to reproduce this issue in your application. I tested in both iOS 15.6 and iOS 16.4 and focus is correctly moved to the modal. Here is a demo on iOS 15.6:
> 
> Does this issue happen if you access the app from a web browser? Or does it only reproduce in Cordova?
> 
> For more context, Ionic will manually move focus inside of the modal when it is presented. We also hide contents underneath the modal using `aria-hidden` so it is not accidentally read by screen readers. We hide contents underneath the modal by adding an `aria-hidden` to the `ion-router-outlet` component. Our playground examples do not use `ion-router-outlet`, which is why focus trapping is not working there. However, your application does use `ion-router-outlet` so content underneath the modal is being hidden.

I have only tested it with Cordova. 

Could the issue be related to delay in rendering or may be an async call? 
In the sample, initially the modal can not be read even with the 'Two-finger swipe up' gesture but if you delete the modal node from DOM and add it back using command+Z, the screen reader will then read the modal content.

**Comment by liamdebeasi** [comment]
Tt seems to have something to do with this element being added to the DOM after the page is loaded (as opposed to the inline modal which is in the DOM on load). It looks like the browser is moving focus to `ion-modal`, but the VoiceOver highlight/outline does not move with it.

**Comment by liamdebeasi** [comment]
It looks like the use of `role="dialog"` and `aria-modal="true"` together triggers this bug in Safari: https://github.com/ionic-team/ionic-framework/blob/451d2204e79a4a10c1eb829ab0bd75c137b02475/core/src/components/modal/modal.tsx#L961-L963

The issue can be avoided by removing one of the attributes. This is an unfortunate WebKit bug that I'm not sure we are able to easily work around inside of Ionic. The good news is that this bug has been fixed since around iOS 15.5. 

`aria-modal="true"` tells screen readers that content outside the modal is inert. However, Ionic manually does this using `aria-hidden` on the router outlet, so you may be able to remove `aria-modal` with no behavior change. The change of course is not ideal, but I'm not sure what the alternative would be other than migrating to inline overlays.

You can do something like this before you call `modal.present()`:

```js
modal.shadowRoot.querySelector('.modal-wrapper').removeAttribute('aria-modal');
```

**Comment by sanaadostmuhammad** [comment]
Thanks @liamdebeasi for looking into it. I will give it a try.

**Comment by sanaadostmuhammad** [comment]
`aria-modal` removal worked. Thanks @liamdebeasi

**Comment by liamdebeasi** [comment]
Glad it worked! I am going to close this, but let me know if you have any questions.
